spark2/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,288 @@
+      startingOffset = new StreamingOffset(startSnapshotId, 0, true, false);
+    } else {
+      List<Long> snapshotIds = SnapshotUtil.currentAncestors(table);
+      if (snapshotIds.isEmpty()) {
+        // there's no snapshot currently.
+        startingOffset = StreamingOffset.START_OFFSET;
+      } else {
+        startingOffset = new StreamingOffset(snapshotIds.get(snapshotIds.size() - 1), 0, true, false);
+      }
+    }
+
+    return startingOffset;
+  }
+
+  private StreamingOffset calculateEndOffset(StreamingOffset start) {
+    if (start.equals(StreamingOffset.START_OFFSET)) {
+      return StreamingOffset.START_OFFSET;
+    }
+
+    // Spark will invoke setOffsetRange more than once. If this is already calulated, use the cached one to avoid
+    // calculating again.
+    if (cachedPendingBatches == null || !cachedPendingBatches.first().equals(start)) {
+      this.cachedPendingBatches = Pair.of(start, getChangesWithRateLimit(start.snapshotId(), start.index(),
+          start.isStartingSnapshotId(), start.isLastIndexOfSnapshot(), maxSizePerBatch));
+    }
+
+    List<MicroBatch> batches = cachedPendingBatches.second();
+    MicroBatch lastBatch = batches.isEmpty() ? null : batches.get(batches.size() - 1);
+
+    if (lastBatch == null) {
+      return start;
+    } else {
+      boolean isStarting = lastBatch.snapshotId() == start.snapshotId() && start.isStartingSnapshotId();
+      return new StreamingOffset(lastBatch.snapshotId(), lastBatch.endFileIndex(), isStarting,
+          lastBatch.lastIndexOfSnapshot());
+    }
+  }
+
+  @VisibleForTesting
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  List<MicroBatch> getChangesWithRateLimit(long snapshotId, int index, boolean isStarting,
+                                           boolean isLastIndexOfSnapshot, long maxSize) {
+    List<MicroBatch> batches = Lists.newArrayList();
+    long currentLeftSize = maxSize;
+    MicroBatch lastBatch = null;
+
+    if (!isLastIndexOfSnapshot && (isStarting || isValidSnapshot(table.snapshot(snapshotId)))) {
+      MicroBatch batch = MicroBatches.from(table.snapshot(snapshotId), table.io())
+          .caseSensitive(super.caseSensitive())
+          .specsById(table.specs())
+          .generate(index, currentLeftSize, isStarting);
+
+      batches.add(batch);
+      currentLeftSize -= batch.sizeInBytes();
+      lastBatch = batch;
+    }
+
+    // Current snapshot can already satisfy the size needs.
+    // Or DataFile size cannot satisfy the current needs of max size. For example: the request left size is 100, but the
+    // file size is 200, then file cannot add to this batch.
+    if (currentLeftSize <= 0L || (lastBatch != null && !lastBatch.lastIndexOfSnapshot())) {
+      return batches;
+    }
+
+    long currentSnapshotId = table.currentSnapshot().snapshotId();
+    if (currentSnapshotId == snapshotId) {
+      // the snapshot of current offset is already the latest snapshot of this table.
+      return batches;
+    }
```

Comment:
I don't think that this case is needed. If the two are equal, then `SnapshotUtil.snapshotIdsBetween` should return an empty list, in which case there are no runs through the loop and the batches are returned.

Reply:
`SnapshotUtil.snapshotIdsBetween` will not return an empty list, instead it will throw an exception when `from` and `to` snapshot is the same.